### ISSUE TITLE
Rule Proposal: Semantic `where`

### ISSUE THREAD

**Issue by aamctustwo** [realm/SwiftLint#281]
**Rule Proposal:** 
A warning should be raised when using `where` in an optional binding or casting pattern (e.g. `if let`, `guard let`, `if let _ = _ as? _`, etc.) but none of the variables in the `where` clause were assigned in the optional binding or casting.

**Triggering Example**
```swift
if let foo = bar as? String where baz != -1 {
  ...
}
```

**Non-Triggering Example**
```swift
if let foo = bar as? String where !foo.isEmpty {
  ....
}
```

**Autocorrection for Triggering Example**
```swift
if baz != -1, let foo = bar as? String {
  ...
}
```

**Discussion**
This provides an early out before more expensive operations, unwrapping/casting and allocating memory, etc. It also improves readability.

**References:**
- http://ericasadun.com/2015/12/18/dear-erica-help-me-refactor/

**Comment by jpsim** [comment]
Thanks for the rule proposal! I agree that this could be useful.

**Comment by scottrhoyt** [comment]
I like this as well. +1

**Comment by proth** [comment]
+1

**Comment by masters3d** [comment]
With swift3 this rule is no longer needed. 
https://github.com/apple/swift-evolution/blob/master/proposals/0099-conditionclauses.md

**Comment by jpsim** [comment]
Thanks for finding that @masters3d! I agree. Closing.
